test/utils/types/Time.test.js

```diff
@@ -0,0 +1,150 @@
+    this.mock = await Time.new();
+  });
+
+  describe('clocks', function () {
+    it('timestamp', async function () {
+      expect(await this.mock.$timestamp()).to.be.bignumber.equal(web3.utils.toBN(await clock.timestamp()));
+    });
+
+    it('block number', async function () {
+      expect(await this.mock.$blockNumber()).to.be.bignumber.equal(web3.utils.toBN(await clock.blocknumber()));
+    });
+  });
+
+  it('isSetAndPast', async function () {
+    for (const timepoint of [0n, 1n, 2n, 17n, 42n, MAX_UINT48])
+      for (const ref of [0n, 1n, 2n, 17n, 42n, MAX_UINT48]) {
+        expect(await this.mock.$isSetAndPast(timepoint, ref)).to.be.equal(timepoint != 0 && timepoint <= ref);
+      }
+  });
+
+  describe('Delay', function () {
+    describe('helpers', function () {
+      const valueBefore = 17n;
+      const valueAfter = 42n;
+      const effect = 69n;
+      const delay = 1272825341158973505578n;
+
+      it('pack', async function () {
+        const packed = await this.mock.$pack(valueBefore, valueAfter, effect);
+        expect(packed).to.be.bignumber.equal(delay.toString());
+
+        const packed2 = packDelay({ valueBefore, valueAfter, effect });
+        expect(packed2).to.be.equal(delay);
+      });
+
+      it('unpack', async function () {
+        const unpacked = await this.mock.$unpack(delay);
+        expect(unpacked[0]).to.be.bignumber.equal(valueBefore.toString());
+        expect(unpacked[1]).to.be.bignumber.equal(valueAfter.toString());
+        expect(unpacked[2]).to.be.bignumber.equal(effect.toString());
+
+        const unpacked2 = unpackDelay(delay);
+        expect(unpacked2).to.be.deep.equal({ valueBefore, valueAfter, effect });
+      });
+    });
+
+    it('toDelay', async function () {
+      for (const value of [0n, 1n, 2n, 17n, 42n, MAX_UINT32]) {
+        const delay = await this.mock.$toDelay(value).then(unpackDelay);
+        expect(delay).to.be.deep.equal({ valueBefore: 0n, valueAfter: value, effect: 0n });
+      }
+    });
+
+    it('getAt & getFullAt', async function () {
+      const valueBefore = 24194n;
+      const valueAfter = 4214143n;
+
+      for (const effect of [0n, 1n, 15n, 16n, 17n, 42n, MAX_UINT48])
+        for (const timepoint of [0n, 1n, 15n, 16n, 17n, 42n, MAX_UINT48]) {
+          const isPast = effect <= timepoint;
+
+          const delay = packDelay({ valueBefore, valueAfter, effect });
+
+          expect(await this.mock.$getAt(delay, timepoint)).to.be.bignumber.equal(
+            String(isPast ? valueAfter : valueBefore),
+          );
+
+          const getFullAt = await this.mock.$getFullAt(delay, timepoint);
+          expect(getFullAt[0]).to.be.bignumber.equal(String(isPast ? valueAfter : valueBefore));
+          expect(getFullAt[1]).to.be.bignumber.equal(String(isPast ? 0n : valueAfter));
+          expect(getFullAt[2]).to.be.bignumber.equal(String(isPast ? 0n : effect));
+        }
+    });
+
+    it('get & getFull', async function () {
```

Comment:
This test relies on the timepoint chosen by the node and may only be testing one of the paths. We should set the chain clock to test what we're interested in, no?

Reply:
The issue with the chain clock is we can never set it in the past, only increase it. That could makes the testing possibly unstable ?

Reply:
I implemented that using a `effectForTimepoint` helper. WDYT ?